profile/gperf.go

```diff
@@ -0,0 +1,132 @@
+package profile
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Common flags of perf
+const (
+	AllCPUs  = "-a"
+	OutFile  = "-o"
+	Detailed = "-d"
+	Event    = "-e"
+	SEP      = "-x"
+)
+
+// PerfStat A instance of perf stat command
+type PerfStat struct {
+	cmd     *exec.Cmd
+	proc    int
+	sep     string
+	outFile string
+}
+
+// NewPerfStat returns a new instance for perf stat
+func NewPerfStat(arguments ...string) *PerfStat {
+	perfStat := new(PerfStat)
+
+	var args, delimiter string
+	for i, arg := range arguments {
+		args += delimiter + arg
+		delimiter = " "
+
+		switch arg {
+		case OutFile:
+			perfStat.outFile = arguments[i+1]
+		case SEP:
+			perfStat.sep = arguments[i+1]
+		}
+	}
+
+	if perfStat.sep == "" {
+		perfStat.sep = "|"
+		args += delimiter + SEP + delimiter + "\\" + perfStat.sep
+	}
+
+	args += delimiter + "&"
+
+	commandString := fmt.Sprintf("perf stat %s", args)
+	perfStat.cmd = exec.Command("/bin/sh", "-c", commandString) // TODO: FIX Perf not running
+	log.Debugf("Perf command: %s", perfStat.cmd)
+
+	return perfStat
+}
+
+// Run runs perf stat command, assumes only one perf process running
+func (p *PerfStat) Run() error {
+	log.Debug("Perf starts")
+	if err := p.cmd.Run(); err != nil {
+		return err
+	}
+
+	// get pid
+	pidBytes, err := exec.Command("pidof", "perf").Output()
+	if err != nil {
+		log.Warnf("Failed to get perf PID: %v", err)
+		return err
+	}
+	pid, err := strconv.Atoi(strings.TrimSpace(strings.Split(string(pidBytes), "\n")[0]))
+	if err != nil {
+		log.Warnf("Pid conversion failed: %v", err)
+		return err
+	}
+
+	p.proc = pid
+
+	time.Sleep(100 * time.Millisecond) // Wait for perf completely start
+	return nil
+}
+
+// Stop stops perf stat command and returns the result in map[event]value
+func (p *PerfStat) Stop() (map[string]float64, error) {
```

Comment:
you should not kill perf. You should run it for a user-specified amount of time during for a user-specified number of rounds.

Reply:
Done

Reply:
I don't get why we need to run perf repeatedly. I thought we need to profile the entire invocations (requests). My implementation is start profiling just after VMs are booted and stop it after all invocations return. If we run perf for multiple rounds, it will just profile partial CPU counters of the whole test. Moreover, I think it is also difficult for users to determine the running time of perf.

Reply:
Done